stable_marriage/gale_shapley.nim

```diff
@@ -0,0 +1,274 @@
+    contenderMatches*: seq[int] ## Matched recipients for each contender
+    recipientMatches*: seq[int] ## Matched contenders for each recipient
+
+  Clash* = tuple[id, match, prefers: Natural]
+
+# Helper functions to prepare the numerical representation of preferences from
+# an array of arrays of names in order of descending preference.
+func initContenderPrefs*[N: static int](prefs: array[N, array[N, string]];
+    rNames: openArray[string]): array[N, array[N, int]] {.compileTime.} =
+  ## Contender's preferences hold the recipient ids in descending order
+  ## of preference.
+  for c, ranking in pairs(prefs):
+    for rank, recipient in pairs(ranking):
+      assert recipient in ranking
+      result[c][rank] = rNames.find(recipient)
+
+func initRecipientPrefs*[N: static int](prefs: array[N, array[N, string]];
+    cNames: openArray[string]): array[N, array[N, int]] {.compileTime.} =
+  ## Recipient's preferences hold the preference score for each contender's id.
+  for r, ranking in pairs(prefs):
+    for rank, contender in pairs(ranking):
+      assert contender in ranking
+      result[r][cNames.find(contender)] = rank
+
+func invertPrefs*[N: static int](prefs: array[N, array[N, int]]):
+                                 array[N, array[N, int]] =
+  ## Converts each element of `prefs` from Ids in order of decreasing
+  ## preference to ranking score for each Id.
+  ## Used to convert from format used for Contenders to one used for Recipients.
+  for rId, ranking in prefs.pairs():
+    for rank, id in ranking.pairs():
+      result[rId][id] = rank
+
+
+func stableMatching*(contenderPrefs, recipientPrefs: openArray[
+    Ranking]): Matches =
+  ## Calculates a stable matching for a given set of preferences.
+  ## Returns an object with corresponding match ids
+  ## for each contender and each recipient.
+  ##
+  ## Each element of the argument arrays is an array of ints, meaning slightly
+  ## different things:
+  ## * `contenderPrefs`: recipient ids in descending order of preference
+  ## * `recipientPrefs`: preference score for each contender's id
+  ##
+  assert recipientPrefs.len == recipientPrefs[0].len
+  assert contenderPrefs.len == contenderPrefs[0].len
+  assert recipientPrefs.len == contenderPrefs.len
+  let rosterLen = recipientPrefs.len
+  var
+    # Initializing result sequences with -1, meaning "unmatched"
+    recMatches = newSeqWith(rosterLen, -1)
+    contMatches = newSeqWith(rosterLen, -1)
+    # Queue holding the currently considered "preference score"
+    # (idx of contenderPrefs) for each contender.
+    contQueue = newSeqWith(rosterLen, 0)
+  template match(c, r: Natural) =
+    # Recipient accepts contender, the match is stored
+    contMatches[c] = r
+    recMatches[r] = c
+  while contMatches.contains(-1): # While exist unmatched contenders...
+    for c in 0..<rosterLen: # for each contender...
+      if contMatches[c] == -1: # if they are yet unmatched...
```

Comment:
So I thought for a bit if I can make it work without this scan-through at the beginning of each big cycle followed by a filter and came up with the following code.

It leverages an additional sequence used as a dequeue for keeping track of unmatched contenders. This allows to drop the `contains` and `if ... == -1` parts, but the trade-off is that the order of traversal changes. Since `del` moves the last element at the beginning of the sequence, this implementation often oscillates between two contenders and scan goes down their corresponding preferences until a match is found for either of them. I suspected this imbalance could increase the number of total iterations until a stable matching for all participants is found.

To test it I pre-generated a random input table for 2048 contenders+2048 recipients and measured the performance of both versions. Surprisingly, the number of iterations turned out to be equal (14874). The difference in execution time is negligible and may even be worse for the new version, and the code is a bit more complex. So, not an improvement overall, so I leave it here just to document that it's been tried.

```nim
func stableMatchingNew*(contenderPrefs, recipientPrefs: openArray[
    Ranking]): Matches =
  let rosterLen = recipientPrefs.len
  var
    # Initializing result sequences with -1, meaning "unmatched"
    recMatches = newSeqWith(rosterLen, -1)
    contMatches = newSeqWith(rosterLen, -1)
    # Queue holding the currently considered "preference score" (idx of
    # contenderPrefs) for each contender, initialized with the most preferred
    contQueue = newSeqWith(rosterLen, 0)
    unmatchedContenders = toSeq(0..<rosterLen) # Queue of vacant contenders
    # A seq is used as a dequeue, traversal of unmatched contenders in unordered
  template match(c, r: Natural) =
    # Recipient accepts contender, the match is stored
    contMatches[c] = r
    recMatches[r] = c
  while unmatchedContenders.len > 0: # While exist unmatched contenders...
    let c = unmatchedContenders[0]
    unmatchedContenders.del(0) # Remove the contender from the vacant list
    let r = contenderPrefs[c][contQueue[c]] # Proposing to the queued recipient
    inc(contQueue[c]) # Queue next preferred recipient for this contender
    assert contQueue[c] <= rosterLen
    let rivalMatch = recMatches[r] # Recipient's match index or -1 (vacant)
    if rivalMatch == -1: # Recipient is also unmatched
      match(c, r)
    # Recipient is matched, but contender is more preferable
    elif recipientPrefs[r][c] < recipientPrefs[r][rivalMatch]:
      contMatches[rivalMatch] = -1 # Vacate current recipient's match and...
      unmatchedContenders.add(rivalMatch) # add them to the list
      match(c, r)
    else: # Contender's proposition is rejected
      unmatchedContenders.add(c)
  Matches(contenderMatches: contMatches, recipientMatches: recMatches)
```

PS: I wonder if there's a difference if tested on some data bearing closer resemblance to the real world, with some participants being evidently more popular and some less so.

Reply:
I promise to have a look at it this weekend. It seems like you hit a strange interesting corner case with this optimisation attempt.
I do not have books as references for the stable marriage problem.
I know some «real-world» applications:
– advertisement for users with preferences in e.g. search engines. (There are maybe better formulations of this problem than stable marriage).
– cryptography, but don't remember exactly where.

Reply:
You need a sorted list of indexes of unmatched contenders.
Since you need the minimal index, a heap would be best suited for this.

Reply:
Is having a *sorted* list a requirement? I haven't thought much about it, but looks like the order doesn't matter much.

Also, my hunch is an overhead of a heap is not worth it for this algorithm. Scanning an array or ints is very cheap, and I think there's a merit in an implementation that doesn't use anything besides arrays/seqs and is still efficient.

Reply:
Sure, it is an overhead for this algorithm. Well, you are right, the order should not matter.